### PR TITLE
Static typing for cells

### PR DESCRIPTION
**Pull request by corbt**
This PR is the first step towards implementing https://github.com/redwoodjs/redwood/issues/2205. It allows users to declare the prop types their cells expect.

It also renames `withCell` to `createCell`, for the reasons I outlined in https://github.com/redwoodjs/redwood/issues/2205#issuecomment-812964440. `createCell` feels like a more parsimonious name to me but happy to back this change out if others disagree.

It appears that the babel transform for wrapping cells already checks for a default export and bails out if one exists, so this shouldn't require any changes there.

The major remaining task is updating the cell generator to add a `export default createCell` line at the end of each cell in Typescript. I'd appreciate more input on whether this is a change we're fine making on both the TS and JS sides, or whether we'll need two templates. Do we have multiple templates anywhere else?

### CODE REVIEW

**Review by peterp — APPROVED** [review]
Looks good to me!

File: packages/core/src/babelPlugins/babel-plugin-redwood-cell.ts [hunk 1]
```diff
@@ -65,24 +65,24 @@ export default function ({ types: t }: { types: typeof types }): PluginObj {
...
                 ),
               ],
               t.stringLiteral('@redwoodjs/web')
             )
           )
 
           // Insert at the bottom of the file:
-          // + export default withCell({ QUERY?, Loading?, Succes?, Failure?, Empty?, beforeQuery?, afterQuery? })
+          // + export default createCell({ QUERY?, Loading?, Succes?, Failure?, Empty?, beforeQuery?, afterQuery? })
           path.node.body.push(
```
Comment (dac09) [inline packages/core/src/babelPlugins/babel-plugin-redwood-cell.ts:83]: Should we have a check here if `createCell` is already declared?

Comment (peterp) [reply packages/core/src/babelPlugins/babel-plugin-redwood-cell.ts:83]: We already abort the automatic Cell creation if there's a default export

**Review by dac09 — APPROVED** [review]

**Review by thedavidprice — CHANGES_REQUESTED** [review]
Hi all! @mojombo has asked us to HODL merging this PR until he's able to dig in later this week / early next. Since he's out this week, he and I decided to not include anything with potential API changes in v0.29 release.

I've added it to the following spring and created a milestone "hold-merge".

Sound good?

**Comment by thedavidprice** [comment]
**Update:** Core Team had a great discussion about this earlier today. @peterp will be following up with the next steps as we'd like to get this into v0.30 (next release). 

We'll loop in Tom one more time once we finalize a the discussion topics from earlier today.

Thanks again!

**Comment by corbt** [comment]
Awesome, will wait to hear from @peterp !

**Review by peterp — APPROVED** [review]

**Comment by peterp** [comment]
I've updated this so that it's no longer a breaking change - we're not 100% confident that people aren't using the `withCell` export, so I've added a deprecation notice that we'll remove it in `v0.31.`.

One of the concerns that we also have about this approach is that if a user declares `beforeQuery` in the cell file, they need to remember to pass that to `createCell`.

I think this may or may-not be a concern, since if the user doesn't export `beforeQuery` they'll get warned about an unused variable. Either way we're going to try and make this feel magical.

File: packages/web/src/index.ts [hunk 2]
```diff
@@ -12,7 +12,11 @@ export {
   useMutation,
 } from './components/GraphQLHooksProvider'
 
-export { withCell } from './components/withCellHOC'
+import { createCell } from './components/createCell'
+export { createCell }
+
+/** @deprecated `withCell` will be removed in v0.31.0, please use `createCell` instead. */
```
Comment (dac09) [inline packages/web/src/index.ts:18]:
```suggestion
/** @deprecated `withCell` will be removed in v0.32.0, please use `createCell` instead. */
```

**Comment by corbt** [comment]
Wohoo! Excited this is in. :)

**Pull request merged**